engine/src/main/java/org/pentaho/di/trans/steps/fuzzymatch/FuzzyMatch.java

```diff
@@ -0,0 +1,629 @@
+          data.nrCachedFields += meta.getValue().length;
+        }
+      }
+      if ( log.isRowLevel() ) {
+        logRowlevel( BaseMessages.getString( PKG, "FuzzyMatch.Log.ReadLookupRow" )
+          + rowSet.getRowMeta().getString( rowData ) );
+      }
+
+      // Look up the keys in the source rows
+      // and store values in cache
+
+      Object[] storeData = new Object[data.nrCachedFields];
+      // Add key field
+      if ( rowData[data.indexOfCachedFields[0]] == null ) {
+        storeData[0] = "";
+      } else {
+        ValueMetaInterface fromStreamRowMeta = rowSet.getRowMeta().getValueMeta( data.indexOfCachedFields[0] );
+        if ( fromStreamRowMeta.isStorageBinaryString() ) {
+          storeData[0] = fromStreamRowMeta.convertToNormalStorageType( rowData[data.indexOfCachedFields[0]] );
+        } else {
+          storeData[0] = rowData[data.indexOfCachedFields[0]];
+        }
+      }
+
+      // Add additional fields?
+      for ( int i = 1; i < data.nrCachedFields; i++ ) {
+        ValueMetaInterface fromStreamRowMeta = rowSet.getRowMeta().getValueMeta( data.indexOfCachedFields[i] );
+        if ( fromStreamRowMeta.isStorageBinaryString() ) {
+          storeData[i] = fromStreamRowMeta.convertToNormalStorageType( rowData[data.indexOfCachedFields[i]] );
+        } else {
+          storeData[i] = rowData[data.indexOfCachedFields[i]];
+        }
+      }
+      if ( isDebug() ) {
+        logDebug( BaseMessages.getString( PKG, "FuzzyMatch.Log.AddingValueToCache", data.infoCache
+          .getString( storeData ) ) );
+      }
+
+      addToCache( storeData );
+
+      rowData = getRowFrom( rowSet );
+
+      if ( firstRun ) {
+        firstRun = false;
+      }
+    }
+
+    return true;
+  }
+
+  private Object[] lookupValues( RowMetaInterface rowMeta, Object[] row ) throws KettleException {
+    if ( first ) {
+      first = false;
+
+      data.outputRowMeta = getInputRowMeta().clone();
+      meta.getFields(
+        data.outputRowMeta, getStepname(), new RowMetaInterface[] { data.infoMeta }, null, this, repository,
+        metaStore );
+
+      // Check lookup field
+      data.indexOfMainField = getInputRowMeta().indexOfValue( environmentSubstitute( meta.getMainStreamField() ) );
+      if ( data.indexOfMainField < 0 ) {
+        // The field is unreachable !
+        throw new KettleException( BaseMessages.getString( PKG, "FuzzyMatch.Exception.CouldnotFindMainField", meta
+          .getMainStreamField() ) );
+      }
+    }
+    Object[] add = null;
+    if ( row[ data.indexOfMainField ] == null ) {
+      add = buildEmptyRow();
+    } else {
+      try {
+        add = getFromCache( row );
+      } catch ( Exception e ) {
+        throw new KettleStepException( e );
+      }
+    }
+    return RowDataUtil.addRowData( row, rowMeta.size(), add );
+  }
+
+  private void addToCache( Object[] value ) throws KettleException {
+    try {
+      data.look.add( value );
+    } catch ( java.lang.OutOfMemoryError o ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)